mono/metadata/android-ifaddrs.c

```diff
@@ -0,0 +1,1158 @@
+		return NULL;
+
+	return iface->ifa_name;
+}
+
+static gint
+get_interface_flags_by_index (gint index, struct _monodroid_ifaddrs **ifaddrs_head)
+{
+	struct _monodroid_ifaddrs *iface = find_interface_by_index (index, ifaddrs_head);
+	if (!iface)
+		return 0;
+
+	return iface->ifa_flags;
+}
+
+static gint
+calculate_address_netmask (struct _monodroid_ifaddrs *ifa, struct ifaddrmsg *net_address)
+{
+	if (ifa->ifa_addr && ifa->ifa_addr->sa_family != AF_UNSPEC && ifa->ifa_addr->sa_family != AF_PACKET) {
+		uint32_t prefix_length = 0;
+		uint32_t data_length = 0;
+		guchar *netmask_data = NULL;
+
+		switch (ifa->ifa_addr->sa_family) {
+			case AF_INET: {
+				struct sockaddr_in *sa = (struct sockaddr_in*)calloc (1, sizeof (struct sockaddr_in));
+				if (!sa)
+					return -1;
+
+				ifa->ifa_netmask = (struct sockaddr*)sa;
+				prefix_length = net_address->ifa_prefixlen;
+				if (prefix_length > 32)
+					prefix_length = 32;
+				data_length = sizeof (sa->sin_addr);
+				netmask_data = (guchar*)&sa->sin_addr;
+				break;
+			}
+
+			case AF_INET6: {
+				struct sockaddr_in6 *sa = (struct sockaddr_in6*)calloc (1, sizeof (struct sockaddr_in6));
+				if (!sa)
+					return -1;
+
+				ifa->ifa_netmask = (struct sockaddr*)sa;
+				prefix_length = net_address->ifa_prefixlen;
+				if (prefix_length > 128)
+					prefix_length = 128;
+				data_length = sizeof (sa->sin6_addr);
+				netmask_data = (guchar*)&sa->sin6_addr;
+				break;
+			}
+		}
+
+		if (ifa->ifa_netmask && netmask_data) {
+			/* Fill the first X bytes with 255 */
+			uint32_t prefix_bytes = prefix_length / 8;
+			uint32_t postfix_bytes;
+			gint i;
+
+			if (prefix_bytes > data_length) {
+				errno = EINVAL;
+				return -1;
+			}
+			postfix_bytes = data_length - prefix_bytes;
+			memset (netmask_data, 0xFF, prefix_bytes);
+			if (postfix_bytes > 0)
+				memset (netmask_data + prefix_bytes + 1, 0x00, postfix_bytes);
+			g_debug ("   calculating netmask, prefix length is %u bits (%u bytes), data length is %u bytes\n", prefix_length, prefix_bytes, data_length);
+			if (prefix_bytes + 2 < data_length)
+				/* Set the rest of the mask bits in the byte following the last 0xFF value */
+				netmask_data [prefix_bytes + 1] = 0xff << (8 - (prefix_length % 8));
+			g_debug ("   netmask is: ");
+			for (i = 0; i < data_length; i++) {
+				g_debug ("%s%u", i == 0 ? " " : ".", (guchar)ifa->ifa_netmask->sa_data [i]);
+			}
+			g_debug ("\n");
+		}
+	}
+
+	return 0;
+}
+
+static struct _monodroid_ifaddrs *
+get_link_address (const struct nlmsghdr *message, struct _monodroid_ifaddrs **ifaddrs_head)
+{
+	size_t length = 0;
+	struct rtattr *attribute;
+	struct ifaddrmsg *net_address;
+	struct _monodroid_ifaddrs *ifa = NULL;
+	struct sockaddr **sa;
+	gint payload_size;
+
+	g_assert (message);
+	net_address = NLMSG_DATA (message);
+	length = IFA_PAYLOAD (message);
+	g_debug ("   address data length: %u", length);
+	if (length <= 0) {
+		goto error;
+	}
+
+	ifa = calloc (1, sizeof (*ifa));
```

Comment:
Might want to cast from void*  to non-void*..

Reply:
Agreed